### PR TITLE
Datadog: default Host to Environment.MachineName

### PR DESCRIPTION
**Pull request by gouri-yerra**
**Motivation**

- Addresses [AGNTLOG-429](https://datadoghq.atlassian.net/browse/AGNTLOG-429) and issue [133](https://github.com/DataDog/serilog-sinks-datadog-logs/issues/133)
- Many users don’t explicitly set Host when configuring the sink. Datadog recommends sending a host to improve grouping and correlation.
- Today, if Host is unset, the payload may lack a host field. This change makes the default behavior safer and simplifies attribution by using a consistent host on all events.

**What’s changing**
Default host resolution:

- If the sink Host option is null/empty, use `Environment.MachineName`.
- On netstandard1.x targets where `Environment.MachineName` isn’t available, fall back to `COMPUTERNAME` or `HOSTNAME` env vars.

Event-level host handling
- The sink always uses the configured/default host. An event-level "host" property is ignored.

Tests:
Added `tests/Serilog.Sinks.Datadog.Logs.Tests/HostTests.cs` to cover:

- Unset Host → defaults to machine name
- Explicit Host respected

**Behavior examples**
Defaulting when not set:

- Sink configured without host → payload includes "host": "<Environment.MachineName>".

Explicit host respected:

- WriteTo.DatadogLogs(host: "web-01") → payload "host": "web-01".

Event-level host ignored

log.ForContext("host", "job-runner-42").Information("...") → payload still uses the configured/default host.

Validation

- Unit tests: 11/11 passing, including new host tests.
- Verified .NET targets build, including netstandard1.3 with the fallback logic.
- Ensured no change to message truncation and existing payload fields.

[AGNTLOG-429]: https://datadoghq.atlassian.net/browse/AGNTLOG-429?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by gouri-yerra** [comment]
Added more tests covering: whitespace/non-string event host ignored, override scoped to single event, and service/tags preserved when overriding.

Added a 0.6.1 entry to CHANGELOG.md describing:
- Default host fallback to machine name (env var fallback on netstandard1.x)
- Event-level "host" override
- New unit tests

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,11 @@
 # Changelog
 
+## 0.6.1
+
+* Default `host` to machine name when sink `Host` option is unset; on `netstandard1.x`, fall back to `COMPUTERNAME`/`HOSTNAME` environment variables.
+* Allow an event-level `"host"` property to override the configured/default host for that event only.
```
Comment (ddrthall) [inline CHANGELOG.md]: Curious about the source request behind this change, can you share it? I like the extensibility and the simplicity of the current change but have worries regarding unexpected behavior/design - want to poke in and see if my worries were already answered in the design process.

File: dotnet-install.sh [hunk 2]
```diff
@@ -0,0 +1,1887 @@
+#!/usr/bin/env bash
```
Comment (ddrthall) [inline dotnet-install.sh]: Was this file included in error?

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs [hunk 3]
```diff
@@ -28,7 +29,8 @@ public DatadogLogRenderer(string source, string service, string host, string[] t
                 new LogEventProperty("ddsource", new ScalarValue(source ?? CSHARP)),
             };
             if (service != null) { props.Add(new LogEventProperty("service", new ScalarValue(service))); }
-            if (host != null) { props.Add(new LogEventProperty("host", new ScalarValue(host))); }
+            var resolvedHost = string.IsNullOrWhiteSpace(host) ? GetDefaultHostName() : host;
```
Comment (ddrthall) [inline src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs]: I think we should lock this new behavior behind a configuration flag, since including unwanted host information would translate to an increase in user costs.

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs [hunk 4]
```diff
@@ -113,5 +149,26 @@ internal string ToDDPayload(string rawPayload)
...
+
+        private static string GetDefaultHostName()
+        {
+#if NETSTANDARD1_0_OR_GREATER && !NETSTANDARD2_0_OR_GREATER
+            // Environment.MachineName is not available on netstandard1.x
+            var fromEnv = Environment.GetEnvironmentVariable("COMPUTERNAME")
+                ?? Environment.GetEnvironmentVariable("HOSTNAME");
+            return string.IsNullOrWhiteSpace(fromEnv) ? null : fromEnv;
+#else
+            return Environment.MachineName;
```
Comment (ddrthall) [inline src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs]: This can throw an exception according the docs, we should handle it gracefully

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs [hunk 5]
```diff
@@ -113,5 +149,26 @@ internal string ToDDPayload(string rawPayload)
 
             return ddPayloadWriter.ToString();
         }
+
+        private static string GetDefaultHostName()
+        {
+#if NETSTANDARD1_0_OR_GREATER && !NETSTANDARD2_0_OR_GREATER
```
Comment (ddrthall) [inline src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs:130]: Nit: Are there any cases where `!NETSTANDARD2_0_OR_GREATER` will allow an Environment.MachineName lookup?  I think `NETSTANDARD1_0_OR_GREATER` might not be needed

File: CHANGELOG.md (hunk 1 above)
Comment (gouri-yerra) [reply CHANGELOG.md]: Hi! Thanks for taking a look. This came from [Feature Request: Get Host from MachineName if not set (#133)](https://github.com/DataDog/serilog-sinks-datadog-logs/issues/133). 
We default the host when unset and keep an event-level override

File: dotnet-install.sh (hunk 2 above)
Comment (gouri-yerra) [reply dotnet-install.sh]: Thanks for catching that. yes, dotnet-install.sh was added by mistake as part of my local setup. It’s not needed in the repo. I’ll remove it from the PR and add it to .gitignore.

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs (hunk 3 above)
Comment (gouri-yerra) [reply src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs]: Ok I gated the host fallback behind a config flag and an env override. I’ve implemented:

- `DatadogConfiguration.ResolveHostIfMissing` (default false).

- In DatadogSink, compute the effective flag from `config.ResolveHostIfMissing` or `DD_LOGS_SINK_RESOLVE_HOST` (accepts 1/true/yes/on), and pass it to `DatadogLogRenderer`.

- In `DatadogLogRenderer`, added an optional `resolveHostIfMissing` constructor arg; when enabled and host is unset, it falls back to machine name; otherwise preserves prior behavior. Event-level "host" still overrides per-event.

- Updated `HostTests.UnsetHostDefaultsToEnvironmentMachineName` to opt-in via `resolveHostIfMissing: true`.

Ran tests to verify:
`dotnet test Serilog.Sinks.Datadog.Logs.sln -v minimal --nologo`

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs (hunk 4 above)
Comment (gouri-yerra) [reply src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs]: Ok! Wrapped host resolution in `GetDefaultHostName()` with try/catch to return null on failure (both env-based on netstandard1.x and `Environment.MachineName` elsewhere).
No behavior change unless `ResolveHostIfMissing` is enabled; in failure cases we now skip adding host rather than throwing.

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs (hunk 5 above)
Comment (gouri-yerra) [reply src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs:130]: only .NET Standard 1.x lacks `Environment.MachineName`. Using just !NETSTANDARD2_0_OR_GREATER would also route net45/net461/net5+/net6+ (where MachineName exists) through the fallback. The combined check keeps the fallback scoped to netstandard1.x.
If you prefer explicitness, I can switch to an explicit list:
```
#if NETSTANDARD1_0 || NETSTANDARD1_1 || NETSTANDARD1_2 || NETSTANDARD1_3 || NETSTANDARD1_4 || NETSTANDARD1_5 || NETSTANDARD1_6
// fallback path
#else
// Environment.MachineName path
#endif
```

**Comment by gouri-yerra** [comment]
Ok after ALP OH, I removed event-level "host" override so events now always use the configured/default host.
Also updated tests and edited CHANGELOG 0.6.1
All tests pass.

**Review by andrewqian2001datadog — COMMENTED** [review]
Hey I think this is looking good, one concern is that I'm not sure if we are actually getting the values from the config. The code [here](https://github.com/DataDog/serilog-sinks-datadog-logs/blob/feat/default-host-machine-name/src/Serilog.Sinks.Datadog.Logs/Configuration/Implementations/Microsoft.Extensions.Configuration/ApplyMicrosoftExtensionsConfiguration.cs#L28-L34) seems to read the values from the config to use, and it does not include the new option `ResolveHostIfMissing`. Let me know if i'm mistaken.

**Comment by gouri-yerra** [comment]
Hey Andrew! Thanks for catching that, you're right. I updated `ApplyMicrosoftExtensionsConfiguration` to propagate the flag from either the DatadogConfiguration instance or bound config section. 
Example use with `appsettings.json`:
```
{
  "Serilog": {
    "WriteTo": [
      {
        "Name": "DatadogLogs",
        "Args": {
          "apiKey": "<redacted>",
          "configuration": {
            "ResolveHostIfMissing": true
          }
        }
      }
    ]
  }
}
```
Lmk if you think I'm missing anything else

File: src/Serilog.Sinks.Datadog.Logs/Configuration/Implementations/Microsoft.Extensions.Configuration/ApplyMicrosoftExtensionsConfiguration.cs [hunk 6]
```diff
@@ -25,13 +25,19 @@ internal static DatadogConfiguration ConfigureDatadogConfiguration(DatadogConfig
...
                 url: datadogConfiguration?.Url ?? section.Url,
                 port: datadogConfiguration?.Port ?? section.Port,
                 useSSL: datadogConfiguration?.UseSSL ?? section.UseSSL,
                 useTCP: datadogConfiguration?.UseTCP ?? section.UseTCP,
                 maxRetries:datadogConfiguration?.MaxRetries ?? section.MaxRetries
             );
+
+            // Ensure we propagate the ResolveHostIfMissing flag from either the provided configuration
+            // instance or the configuration section binding.
+            result.ResolveHostIfMissing = datadogConfiguration?.ResolveHostIfMissing ?? section.ResolveHostIfMissing;
```
Comment (ddrthall) [inline src/Serilog.Sinks.Datadog.Logs/Configuration/Implementations/Microsoft.Extensions.Configuration/ApplyMicrosoftExtensionsConfiguration.cs]: Nit: having this in the constructor to mirror the other configuration options feels slightly cleaner, unless there's a reason it's split out.

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs [hunk 7]
```diff
@@ -20,15 +21,22 @@ public class DatadogLogRenderer
...
-        public DatadogLogRenderer(string source, string service, string host, string[] tags, int maxMessageSize, ITextFormatter formatter, int? ddPayloadSize = null)
+        public DatadogLogRenderer(string source, string service, string host, string[] tags, int maxMessageSize, ITextFormatter formatter, int? ddPayloadSize = null, bool resolveHostIfMissing = false)
         {
 
             var props = new List<LogEventProperty> {
                 new LogEventProperty("ddsource", new ScalarValue(source ?? CSHARP)),
             };
             if (service != null) { props.Add(new LogEventProperty("service", new ScalarValue(service))); }
-            if (host != null) { props.Add(new LogEventProperty("host", new ScalarValue(host))); }
+            _resolveHostIfMissing = resolveHostIfMissing;
```
Comment (ddrthall) [inline src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs]: Nit: Is _resolveHostIfMissing used outside of this function, or can we just utilize the base resolveHostIfMissing (no underscore) everywhere

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs [hunk 8]
```diff
@@ -89,6 +99,11 @@ internal IEnumerable<byte[]> TruncateIfNeeded(string rawPayload)
         }
 
         internal string ToDDPayload(string rawPayload)
+        {
+            return ToDDPayload(rawPayload, _props);
```
Comment (ddrthall) [inline src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs:99]: Nit: Is the _props change still required? I think it might be leftover from the older change

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs [hunk 9]
```diff
@@ -113,5 +128,40 @@ internal string ToDDPayload(string rawPayload)
...
+                return Environment.MachineName;
+            }
+            catch
+            {
+                return null;
+            }
+#endif
+        }
+
+        private static string TryConvertScalarToString(LogEventPropertyValue value)
```
Comment (ddrthall) [inline src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs]: This function looks like it's safe to remove

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,5 +1,11 @@
 # Changelog
 
+## 0.6.1
+
+* Default `host` to machine name when sink `Host` option is unset; on `netstandard1.x`, fall back to `COMPUTERNAME`/`HOSTNAME` environment variables.
+* Always use the configured/default host for events; event-level `"host"` property is ignored.
```
Comment (ddrthall) [inline CHANGELOG.md]: Nit: Is this a functionality change from before? Or have we always used configured host for events, ignoring event-level host properties?

**Review by ddrthall — DISMISSED** [review]
Left some nits, but the core behavior seems solid to me

File: src/Serilog.Sinks.Datadog.Logs/Configuration/Implementations/Microsoft.Extensions.Configuration/ApplyMicrosoftExtensionsConfiguration.cs (hunk 6 above)
Comment (gouri-yerra) [reply src/Serilog.Sinks.Datadog.Logs/Configuration/Implementations/Microsoft.Extensions.Configuration/ApplyMicrosoftExtensionsConfiguration.cs]: Ok I looked through where `DatadogConfiguration` is ref so I can move `ResolveHostIfMissing` into it. Binder uses constructor instead of post assignment!
Constructor now looks like:
DatadogConfiguration.cs Lines 12-18
```
public bool ResolveHostIfMissing { get; set; }
// ...
public DatadogConfiguration(string url = DDUrl, int port = DDPort, bool useSSL = true, bool useTCP = false, int maxRetries = 10, bool resolveHostIfMissing = false)
{
    Url = url;
    Port = port;
    UseSSL = useSSL;
    UseTCP = useTCP;
    MaxRetries = maxRetries;
    ResolveHostIfMissing = resolveHostIfMissing;
}
```

And config binding passes it via the constructor:
ApplyMicrosoftExtensionsConfiguration.cs Lines 28-34
```
var result = new DatadogConfiguration(
    url: datadogConfiguration?.Url ?? section.Url,
    port: datadogConfiguration?.Port ?? section.Port,
    useSSL: datadogConfiguration?.UseSSL ?? section.UseSSL,
    useTCP: datadogConfiguration?.UseTCP ?? section.UseTCP,
    maxRetries: datadogConfiguration?.MaxRetries ?? section.MaxRetries,
    resolveHostIfMissing: datadogConfiguration?.ResolveHostIfMissing ?? section.ResolveHostIfMissing
);
```

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs (hunk 7 above)
Comment (gouri-yerra) [reply src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs]: Oooh. Yeah it was only used inside the constructor. I removed it so we can just use `resolveHostIfMissing`
DatadogLogRenderer.cs Lines 24-41
```
// private readonly bool _resolveHostIfMissing;  // removed

public DatadogLogRenderer(..., bool resolveHostIfMissing = false)
{
    var props = new List<LogEventProperty> {
        new LogEventProperty("ddsource", new ScalarValue(source ?? CSHARP)),
    };
    if (service != null) { props.Add(new LogEventProperty("service", new ScalarValue(service))); }
    string resolvedHost = host;
    if (string.IsNullOrWhiteSpace(resolvedHost) && resolveHostIfMissing)
    {
        resolvedHost = GetDefaultHostName();
    }
    if (resolvedHost != null) { props.Add(new LogEventProperty("host", new ScalarValue(resolvedHost))); }
    if (tags != null) { props.Add(new LogEventProperty("ddtags", new ScalarValue(string.Join(",", tags)))); }
    _props = props;
    _maxMessageSize = maxMessageSize;
    _formatter = formatter;
    _ddPayloadSize = ddPayloadSize ?? Encoding.UTF8.GetByteCount(ToDDPayload(""));
}
```

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs (hunk 8 above)
Comment (gouri-yerra) [reply src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs:99]: Yeah it was leftover! Removed!
`DatadogLogRenderer.cs` Lines 50-63
```
return TruncateIfNeeded(rawPayload)
    .Select(x => ToDDPayload(Encoding.UTF8.GetString(x)))
    .ToArray();
```

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs (hunk 9 above)
Comment (gouri-yerra) [reply src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogLogRenderer.cs]: Thanks for helping me clean up! removed `TryConvertScalarToString` from `DatadogLogRenderer`

File: CHANGELOG.md (hunk 10 above)
Comment (gouri-yerra) [reply CHANGELOG.md]: Not a functionality chang, we’ve always used the configured/default host for the top-level field and not overridden it from event properties. I clarified this in the `changelog`. I reworded the note to avoid implying a change.

I'm trying to answer this question: 

> In the event that the host variable is not set, can this use `Environment.MachineName` (since that's where the MachineName Enricher gets it from)?

So when host isn’t provided and `ResolveHostIfMissing` is enabled (via config or `DD_LOGS_SINK_RESOLVE_HOST`), we resolve it to `Environment.MachineName`. 

On `netstandard1.x`, where `Environment.MachineName` isn’t available, we fall back to `COMPUTERNAME`/`HOSTNAME` env vars. This mirrors the MachineName enricher’s source.

**Comment by gouri-yerra** [comment]
Thanks for the reviews ! all the points:

- Added `ResolveHostIfMissing` to `DatadogConfiguration` constructor and bound it via `ApplyMicrosoftExtensionsConfiguration` (consistent with other options). Also respected the flag from config or env `DD_LOGS_SINK_RESOLVE_HOST`.
- Simplified `DatadogLogRenderer`: removed the `_resolveHostIfMissing` field and the temporary `propsToUse` variable; use the constructor parameter and `_props` directly. Removed the unused `TryConvertScalarToString`.
- TLDR: Host resolution: when host isn’t provided and `ResolveHostIfMissing` is enabled, we now resolve to `Environment.MachineName`; on netstandard1.x, fall back to `COMPUTERNAME`/`HOSTNAME`.
- Behavior clarification (no change): event-level "host" does not override the configured/default host. Updated `CHANGELOG.md`.

File: CHANGELOG.md (hunk 10 above)
Comment (ddrthall) [reply CHANGELOG.md]: Nit x2: If we're not providing a functional difference then we can safely keep it out of the changelog

**Review by ddrthall — DISMISSED** [review]

File: CHANGELOG.md (hunk 10 above)
Comment (gouri-yerra) [reply CHANGELOG.md]: Ok! Deleted the section that had ResolveHostIfMissing and the MachineName fallback.

**Review by ddrthall — APPROVED** [review]
